tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html

```diff
@@ -201,35 +201,65 @@
               </paper-listbox>
             </paper-dropdown-menu>
           </div>
-          <div class="section paging">
-            <div class="section-title">Paging</div>
-            <div>
-              Number of matching session groups: [[_totalSessionGroupsCountStr]]
-            </div>
-            <div class="inline-element page-number-input">
-              <paper-input
-                label="Page #"
-                value="{{_pageNumberInput.value}}"
-                allowed-pattern="[0-9]"
-                error-message="Invalid input"
-                invalid="[[_pageNumberInput.invalid]]"
-                on-value-changed="_queryServer"
-              >
-                <div slot="suffix" class="page-suffix">/ [[_pageCountStr]]</div>
-              </paper-input>
+          <vaadin-split-layout vertical id="paging-download">
+            <div class="section paging">
+              <div class="section-title">Paging</div>
+              <div>
+                Number of matching session groups:
+                [[_totalSessionGroupsCountStr]]
+              </div>
+              <div class="inline-element page-number-input">
+                <paper-input
+                  label="Page #"
+                  value="{{_pageNumberInput.value}}"
+                  allowed-pattern="[0-9]"
+                  error-message="Invalid input"
+                  invalid="[[_pageNumberInput.invalid]]"
+                  on-value-changed="_queryServer"
+                >
+                  <div slot="suffix" class="page-suffix">
+                    / [[_pageCountStr]]
+                  </div>
+                </paper-input>
+              </div>
+              <div class="inline-element page-size-input">
+                <paper-input
+                  label="Max # of session groups per page:"
+                  value="{{_pageSizeInput.value}}"
+                  allowed-pattern="[0-9]"
+                  error-message="Invalid input"
+                  invalid="[[_pageSizeInput.invalid]]"
+                  on-value-changed="_queryServer"
+                >
+                </paper-input>
+              </div>
             </div>
-            <div class="inline-element page-size-input">
-              <paper-input
-                label="Max # of session groups per page:"
-                value="{{_pageSizeInput.value}}"
-                allowed-pattern="[0-9]"
-                error-message="Invalid input"
-                invalid="[[_pageSizeInput.invalid]]"
-                on-value-changed="_queryServer"
-              >
-              </paper-input>
+            <div class="section download">
+              <template is="dom-if" if="[[_sessionGroupsRequest]]">
+                Download data as
+                <span>
+                  <a
+                    id="csvLink"
+                    download="hparams_table.csv"
+                    href="[[_csvUrl(_sessionGroupsRequest, _hparams, _metrics)]]"
```

Comment:
These deps need to be on `_hparams.*` and `_metrics.*` (as in the
observer to `_updateConfiguration` above) for the events to listen
properly. As written, I can get the frontend into states where the table
view columns don’t match the columns in the downloaded table.

But simpler, I think, is to depend directly on `configuration` and pull
the visibility off of `configuration.columnsVisibility`. That way, you
don’t need to recompute it, and you’re depending on what you actually
need rather than just one of its dependents.

This patch works for me:

```diff
diff --git a/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html b/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html
index 102adaee..55589ecd 100644
--- a/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html
+++ b/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html
@@ -241,19 +241,19 @@ TODO(erez): Add aggregation controls for repeated sessions.
                   <a
                     id="csvLink"
                     download="hparams_table.csv"
-                    href="[[_csvUrl(_sessionGroupsRequest, _hparams, _metrics)]]"
+                    href="[[_csvUrl(_sessionGroupsRequest, configuration)]]"
                     >CSV</a
                   >
                   <a
                     id="jsonLink"
                     download="hparams_table.json"
-                    href="[[_jsonUrl(_sessionGroupsRequest, _hparams, _metrics)]]"
+                    href="[[_jsonUrl(_sessionGroupsRequest, configuration)]]"
                     >JSON</a
                   >
                   <a
                     id="latexLink"
                     download="hparams_table.tex"
-                    href="[[_latexUrl(_sessionGroupsRequest, _hparams, _metrics)]]"
+                    href="[[_latexUrl(_sessionGroupsRequest, configuration)]]"
                     >LaTeX</a
                   >
                 </span>
@@ -589,17 +589,17 @@ TODO(erez): Add aggregation controls for repeated sessions.
         this._queryServer();
       },
 
-      _csvUrl(request, _, __) {
-        return this._downloadDataUrl(request, 'csv');
+      _csvUrl(request, configuration) {
+        return this._downloadDataUrl(request, configuration, 'csv');
       },
-      _jsonUrl(request, _, __) {
-        return this._downloadDataUrl(request, 'json');
+      _jsonUrl(request, configuration) {
+        return this._downloadDataUrl(request, configuration, 'json');
       },
-      _latexUrl(request, _, __) {
-        return this._downloadDataUrl(request, 'latex');
+      _latexUrl(request, configuration) {
+        return this._downloadDataUrl(request, configuration, 'latex');
       },
-      _downloadDataUrl(request, format) {
-        const visibility = this._computeColumnsVisibility();
+      _downloadDataUrl(request, configuration, format) {
+        const visibility = configuration.columnsVisibility;
         return this.backend.getDownloadUrl(format, request, visibility);
       },
 
```

Reply:
Thanks for the suggestion, much better :)